### ISSUE TITLE
Highlight text and images on a page and make annotations

### ISSUE THREAD

**Issue by kaidatavis** [Vis4Sense/SenseMapExtension#144]

**Comment by Yuu-Han** [comment]
Update: c31b896c74797e98e960b15d3f1d8e2d1608a0e0 supports highlighting selected text on the tab page (reuse the implementation in old-historymap with a few apis migrated to manifest v3 features). The highlighted content is not sent to HistoryMap yet.

**Comment by Yuu-Han** [comment]
I just had a look at the wiki page and found the comments on [missing features](https://github.com/Vis4Sense/HistoryMap/wiki/Missing-Features-for-Online-Study).

> Image highlighting * (what about multiple images?)

I think highlighting multiple images is a practical need. For example, when I searching for inspiring prompts and images on a web gallery, I would like to highlight all the images that might be desirable.

The original implementation allows only one representative image by removing any previously highlighted image when there is a new one (see the following code). It should be easy to push multiple images to `highlights` array, the main question is how to present the images in HistoryMap. The options can be: (1) gallery (like mac finder), (2) list.

https://github.com/Vis4Sense/HistoryMap/blob/c31b896c74797e98e960b15d3f1d8e2d1608a0e0/old-historymap/src/background/contentScriptData.js#L15-L30

**Comment by kaidatavis** [comment]
> the main question is how to present the images in HistoryMap.

Indeed, and it needs to be compact so it would still work when there are many nodes and images. I don't have a strong preference between the two; maybe it also relates to the layout, i.e., how the node size/shape with images would better work with the layout.

**Comment by Yuu-Han** [comment]
Implemented highlighting images and showing the highlighted text and images on historymap (902f4ab7e3790afea7810a841b9ac6bbf6690646).

- Images are saved as base64 data stored in the `highlights` array in `hmPages`.

A preview figure:

![image](https://github.com/Vis4Sense/HistoryMap/assets/72588097/355c5ffb-2056-487a-8838-3c9d74a8ba4c)

Todo

- Removing image and selecting representative image is not supported yet
- Currently multiple images are shown as gallery, I might take the layout into consideration and investigate other options.
- Better to save highlights in background

**Comment by Yuu-Han** [comment]
Highlight text is back (252fce674811c6c612267ba66d4cc48459b553fc).

Take note on a webpage (6b14a0a6da4928f01ac9111aa6bcfe3e1ced9675):

![hm-add-note](https://github.com/Vis4Sense/HistoryMap/assets/72588097/d26f9fb5-9856-496f-811d-625f42e21389)

Loading highlights and note from historymap to a reopened webpage is not yet supported.

**Comment by Yuu-Han** [comment]
Editing note on HistoryMap (5edec7e003eea3ff478bee58bc460c90f87310f4):

![image](https://github.com/Vis4Sense/HistoryMap/assets/72588097/d56ed502-97a0-4390-82cd-25f06e6ac818)

**Comment by kaidatavis** [comment]
Add annotation and tag in the same window/box. Adding a hashtag # marks it is a tag.
